### PR TITLE
added portfolio endpoints to create a portfolio page

### PR DESCRIPTION
**Pull request by KarimKhalil33**
## 📝 Description

This PR implements a comprehensive Portfolio API endpoint that provides rich project data for dashboard visualization and portfolio generation. 
The implementation includes: 
New Features: 
- GET /api/portfolio - Returns comprehensive portfolio data with optional project filtering via query parameters 
- GET /api/portfolio-dashboard - Serves the portfolio dashboard HTML interface 
- GET portfolio.js - Serves the dashboard JavaScript file Key Capabilities: Project filtering with comma-separated project IDs (?project_ids=id1,id2,id3) Rich data structure including overview stats, project metrics, skills timeline, language distribution, complexity analysis, and monthly activity Comprehensive error handling with detailed HTTP status codes Custom response headers for client metadata (X-Portfolio-Projects, X-Portfolio-Generated, X-Portfolio-Filtered) Support for both GitHub and local project analysis

> Please include a summary of the change and which issue is fixed.  
> Include relevant motivation and context.  
> List any dependencies that are required for this change.

**Closes:** #555 

---

## 🔧 Type of Change

- [ ] 🐛 Bug fix (non-breaking change that fixes an issue)
- [X] ✨ New feature (non-breaking change that adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] 📚 Documentation added/updated
- [X] ✅ Test added/updated
- [X] ♻️ Refactoring
- [ ] ⚡ Performance improvement

---

## 🧪 Testing

> Please describe how you tested this PR (both manually and with tests).  
> Provide instructions so we can reproduce.

- [x] GET [/api/portfolio] without filters (all projects)
- [x] GET [/api/portfolio]with specific project IDs
- [x] GET [/api/portfolio] with whitespace handling in parameters
- [x] GET [/api/portfolio] with empty project IDs
- [x] Database error handling
- [x] Response headers validation

---

## ✓ Checklist

- [x] 🤖 GenAI was used in generating the code and I have performed a self-review of my own code
- [x] 💬 I have commented my code where needed
- [x] 📖 I have made corresponding changes to the documentation
- [x] ⚠️ My changes generate no new warnings
- [x] ✅ I have added tests that prove my fix is effective or that my feature works and tests are passing locally
- [x] 🔗 Any dependent changes have been merged and published in downstream modules
- [x] 📱 Any UI changes have been checked to work on desktop, tablet, and/or mobile

---

## 📸 Screenshots

> If applicable, add screenshots to help explain your changes
Screenshot 2026-01-29 at 10 54 14 PM

<details>
<summary>Click to expand screenshots</summary>

</details>

### CODE REVIEW

File: app/utils/generate_portfolio.py [hunk 1]
```diff
@@ -176,6 +207,132 @@ def get_skills_timeline(cursor: sqlite3.Cursor, project_ids: Optional[List[str]]
...
+        "project_sizes": project_sizes,
+        "avg_size": sum(p["lines"] for p in project_sizes) / len(project_sizes) if project_sizes else 0
+    }
+
+def get_score_distribution(cursor: sqlite3.Cursor, project_ids: Optional[List[str]] = None) -> Dict[str, Any]:
+    """Get project score distribution for quality analysis."""
+    if project_ids:
+        placeholders = ",".join(["?"] * len(project_ids))
+        cursor.execute(f"""
+            SELECT project_signature, rank
```
Comment (abstractafua) [inline app/utils/generate_portfolio.py]: Just flagging that if PR #554  gets merged before this one we may have to change rank --> score for all DB retrievals

**Review by abstractafua — APPROVED** [review]

File: tests/api/test_portfolio.py [hunk 2]
```diff
@@ -1,7 +1,7 @@
 import pytest
```
Comment (6s-1) [inline tests/api/test_portfolio.py]: Really appreciate the expanded test coverage for the portfolio endpoints — especially the validation of response structure, headers, and edge cases like empty or malformed project_ids. This significantly improves confidence in the API behavior.

File: app/api/routes/portfolio.py [hunk 3]
```diff
@@ -1,36 +1,46 @@
 from typing import Optional, List
```
[the remark this answers is not in the document]
Comment (KarimKhalil33) [reply app/api/routes/portfolio.py]: Thanks for the comment shreya! i already removed the POST portfolio endpoint and replaced it with a get since it made more sense for our application logic.
